### PR TITLE
FilterHelper.py: Move filter code to FilterHelper

### LINKED ISSUE #4494: Move @filter code into FilterHelper
This avoids annotating filter functions with `@filter`. One file less to maintain.

### PR DESCRIPTION
**Pull request by atultherajput**
This avoids annotating filter functions with @filter.

Closes #4494

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/parsing/FilterHelper.py [hunk 1]
```diff
@@ -14,7 +15,21 @@ def is_valid_filter(filter):
     return filter in available_filters
 
 
-def apply_filter(filter_name, filter_args, all_bears=None):
+def apply_filter(filter_name, filter_args, all_bears=None, filter_function):
```
Comment (ayush1999) [inline coalib/parsing/FilterHelper.py]: The order of arguments in wrong. Default arguments are always present after non-default arguments. The correct order should be: `filter_name, filter_args, filter_function, all_bears=None`

**Comment by jayvdb** [comment]
Removing from review queue as we have two PRs for the same issue.  See the issue.

File: coalib/parsing/FilterHelper.py (hunk 1 above)
Comment (shreyans800755) [reply coalib/parsing/FilterHelper.py]: If we are changing the API, I think its better to keep it like this:
`filter_function, filter_name, filter_args, all_bears=None`.

@Makman2 Can we change this API ?

File: coalib/parsing/FilterHelper.py [hunk 2]
```diff
@@ -14,7 +15,21 @@ def is_valid_filter(filter):
...
+    def filter_section_bears(bears, args):
+        return {section:
+                tuple(bear for bear in bears[section]
+                      if filter_function(bear, args))
+                for section in bears}
+
+    @wraps(filter_function)
+    def filter_wrapper(section_to_bears_dict, args):
```
Comment (shreyans800755) [inline coalib/parsing/FilterHelper.py]: We don't this wrapper anymore. It was used to create annotation. Please inline this as well.

File: coalib/parsing/FilterHelper.py (hunk 1 above)
Comment (Makman2) [reply coalib/parsing/FilterHelper.py]: If it was not part of a release yet it's okay :+1:

However, is `filter_function` necessary? This doesn't look very good and seems to defy the design^^

Comment (shreyans800755) [reply coalib/parsing/FilterHelper.py]: Yup, not required. It should be derived from `filter_name`.

**Comment by shreyans800755** [comment]
@atultherajput 
There's some refactoring required in the changes. But before that, your code changes are not passing the tests.
Sadly, `--debug` is not working with `--show-bears` option. There's already some work going on at some other PR. So, if you're not very familier with current testing framework, you can directly fire coala commands which are present in the tests to trace where is it going wrong. You can use pdb as well to debug these issues.

Also, update your shortlog to remove `@filter`. It is actually creating a link to some github user called `filter`. So, at lease remove `@` first.

**Comment by atultherajput** [comment]
@shreyans800755 Well I update the commit message. And currently facing a situation where my OS(Ubuntu) stuck at boot splash screen after upgrading from 17.04 to 17.10. So I cannot access my development environment. Any idea how to resolve it?

**Comment by shreyans800755** [comment]
Oh :disappointed: 
You might wanna try bootable ubuntu to fix that. Just install ubuntu in a pen drive/CD as bootable and try to fix the issue. It usually works for me.

**Comment by atultherajput** [comment]
@shreyans800755 
I tried but still I am not able to debug the error. What to do now? :(

**Comment by shreyans800755** [comment]
@atultherajput Replied on gitter.

**Comment by Nosferatul** [comment]
@atultherajput seems like the bear return with error. can you try to debug locally?

**Comment by Makman2** [comment]
Closing due to inactivity.

**Pull request closed without merging**